### ISSUE TITLE
Consider backing HyList with a deque instead of a list

### ISSUE THREAD

**Issue by gilch** [hylang/hy#1406]
Most Lisps use a singly-linked list as their code data structure. These are optimized for a push/pop at the *head* of the structure. Python's lists are optimized at the *tail*.

Why do Lisp's linked lists cons like `(1 . (2 . (3 . nil)))` when they could have just as easily done the reverse, like `(((nil . 1) . 2) . 3)`? I think it's due to Lisp's prefix notation. It's easier to compile a list from left to right. Looking at Hy's compiler bears this out--notice all the `expr.pop(0)` calls. I counted 99 of them. You're really supposed to avoid doing that in a Python list, since it has to shift down all the remaining elements.

Lisp macros will also often build expressions by left-appending elements with cons. Right-appending is also fairly common, but this is still pretty easy with unquote-splice/dotted lists, and recursion. If Hy's compound models (`HyList` and derivatives) were based on Python's `collections.queue`, insertion and deletion would be easier at the head, and still available at the tail, since deques have most of the list methods. This would be helpful in macros.

The disadvantage is that deques don't implement slicing, and while it supports the random access methods, this is slow except near either end. This is because CPython's deques are backed by a doubly-linked list of blocks, instead of a simple gap buffer of pointers like you might expect. Anything that uses slicing will have to be rewritten. But, looking at the compiler and macros that do this, slicing is mostly used as a substitute for `popleft` (or `pop`) anyway, again better done with a deque. For those rare times you *really* need a Python list in a macro (for random access or slicing), you can use one, then use it to construct the final `HyExpression` (or whatever) expansion at the end.

**Comment by Kodiologist** [comment]
To what end, for speed? I doubt that compilation times for real code would benefit much from this.

**Comment by gilch** [comment]
> real code

I expect that directly human-written code won't be too long, but aren't you underestimating macroexpansions? Especially when macros expand to other macros. They could be several pages long. *That's* what the compiler has to chew through. It adds up.

> To what end, for speed?

It's *really* hard to judge that without profiling. I have no idea where the bottleneck in the compiler is. I agree, early optimization *at the cost of clarity* is an evil. Compilation speed is also a lot less important than runtime speed. But since we seem to be using them as deques *anyway*, wouldn't they be *clearer* in both the compiler and macros if they were really deques? Woudn't you rather write `(.popleft foo)` instead of `(setv foo (cut foo 1))` in your macros?

But a speedup wouldn't surprise me. On a modern system, non-I/O bottlenecks are *typically* cache misses. CPython's `deque` takes this into account, and is better at being a deque than a list is.

Do we ever use random access in the middle of a HyExpression? What about the other compound Hy models? Maybe HyList should stay a list and only HyExpression should be a deque? This is a bit more like how Clojure works.

We could use `iexpr = iter(expr)` and `next(iexpr)` in the compiler instead of `expr.pop(0)` when we want to pop from the front. I've done similar things in macros. But this is an extra step we wouldn't need if they were deques to begin with.

**Comment by Kodiologist** [comment]
> Woudn't you rather write `(.popleft foo)` instead of `(setv foo (cut foo 1))` in your macros?

Usually, I use `(defn shift [l] (.pop l 0))`. (The name `shift` for this is from Perl.) But on Python 3, `(setv [a b #* the-rest] foo)` works too.

**Comment by kirbyfan64** [comment]
I mean, as much as I don't see a problem with this proposal, I'd also like to see concrete evidence that this actually helps. Yes, `x.pop(0)` is bad, but Python stores lists internally as a bunch of pointers. Expressions are pretty much *never* over 10 elements, so an `expr.pop(0)` would end up shifting around 80 bytes. CPython's pop delegates to [list_ass_slice](https://github.com/python/cpython/blob/master/Objects/listobject.c#L580) (don't ask about the name...), which uses the `mem*` operations. glibc's memcpy uses careful prefetching to move roughly 64 bytes at a time (8 per instruction, for 8 instructions, but the memory's already been prefetched).

So TL;DR: Literally everything else in the Hy compiler, as well as in Python itself, probably takes more time than the `expr.pop(0)` calls. Expressions aren't long enough for a shift to actually be significantly slow, and we'd probably burn through more time with the direct indexing now having to go through multiple blocks.

**Comment by gilch** [comment]
> (this basically means `x.pop(0)` is the same thing as `x[:] = x[1:]`

Thanks, that's good to know.

> Expressions are pretty much never over 10 elements

I dispute this. My destructure macros #1328 can easily make `setv` expressions longer than this with its intended typical use. And it could get used a lot. The compiler has to worry about *arbitrary macroexpansions*, not just expressions short enough for humans to type. And that means macros themselves would be using expressions that long.

> we'd probably burn through more time with the direct indexing now having to go through multiple blocks

[How big are these blocks?](https://hg.python.org/cpython/file/3.5/Modules/_collectionsmodule.c). I thought they were 64. For your typical 10-element expressions, wouldn't the lookup performance be about the same as a list?

How often do we even use direct indexing into the middle of a `HyExpression`? Most of the direct indexing I've seen is at the beginning or the end, which a deque would handle just fine.

> I'd also like to see concrete evidence that this actually helps

It's status quo bias maybe. Getting enough evidence to satisfy you might be more effort than changing the implementation. What if the status quo was reversed? Imagine, if they had been backed by deques in the first place, how would you feel about changing them to lists? What about with the argument that the performance hit wouldn't matter, because it's not the bottleneck anyway? Why? This is **not** just about speed. It's about providing the *right interface* for the use case. Left-appending to a code list is a typical use case in Lisp. That's all a cons list can do well.

One thing to be said about the status quo: breaking changes are not good. I don't see this proposal as being as important as some other things we're working on. But the time for any breaking change is sooner, rather than later.

**Comment by Kodiologist** [comment]
> This is **not** just about speed. It's about providing the *right interface* for the use case.

But the interface is distinct from the implementation, right? For example, we can give HyExpression a method `popleft` even if it's a list internally.

**Comment by gilch** [comment]
> the interface is distinct from the implementation

Very true. If the interface is settled, then changing the implementation does not break API, and we're free to do this later, even after 1.0. I think it would be easier to add list methods to a `deque` subclass  via `abc.MutableSequence`, than the other way around, but I'm not sure. With the right methods, we could implement slicing for a `deque`. No existing code would *have* to change if the old list methods are still available and work the same way.

**Comment by kirbyfan64** [comment]
I mean... I'm still skeptical. I could try moving it over, then benchmark the compiler on some typical programs and see if there's even a difference.

**Comment by Kodiologist** [comment]
Come to think of it, this proposal is basically obsolete since the change to make Hy models immutable.
